parachain/pallets/basic-channel/src/lib.rs

```diff
@@ -0,0 +1,115 @@
+//! # Basic Channel
+//!
+//! The Basic Channel module is is a non-incentivized bridge between Ethereum and Pokadot
+//! ecosystems. It is meant to be a low-barrier entry to cross-chain communication without
+//! guarantees, but also without fees.
+//!
+//! ## Implementation
+//!
+//! Before a [`Message`] is dispatched to a target [`Application`], it is submitted to a [`Verifier`] for verification.
+//!
+//! ## Interface
+//!
+//! ### Dispatchable Calls
+//!
+//! - `submit`: Submit a message for verification and dispatch.
+//!
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{decl_error, decl_event, decl_module, decl_storage, dispatch::DispatchResult};
+use frame_system::{self as system, ensure_signed};
+
+use artemis_core::{
+	Message, MessageCommitment, MessageDispatch, MessageId, SubmitOutbound, Verifier,
+};
+use channel::{inbound::BasicInboundChannel, outbound::BasicOutboundChannel};
+use envelope::Envelope;
+use primitives::{InboundChannel, InboundChannelData, OutboundChannel, OutboundChannelData};
+use sp_core::H160 as EthAddress;
+use sp_std::convert::TryFrom;
+use sp_std::prelude::*;
+
+mod channel;
+mod envelope;
+#[cfg(test)]
+mod mock;
+mod primitives;
+#[cfg(test)]
+mod tests;
+
+type MessageNonce = u64;
+
+pub trait Config: system::Config {
+	type Event: From<Event<Self>> + Into<<Self as system::Config>::Event>;
+
+	/// Verifier module for message verification.
+	type Verifier: Verifier;
+
+	/// Used by outbound channels to persist messages for outbound delivery.
+	type MessageCommitment: MessageCommitment;
+
+	/// Verifier module for message verification.
+	type MessageDispatch: MessageDispatch<MessageId>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Config> as BasicChannelModule {
+		/// Storage for inbound channels.
+		pub InboundChannels: map hasher(identity) EthAddress => InboundChannelData;
+		/// Storage for outbound channels.
+		pub OutboundChannels: map hasher(identity) T::AccountId => OutboundChannelData;
+	}
+}
+
+decl_event!(
+	pub enum Event<T>
+	where
+		AccountId = <T as system::Config>::AccountId,
+	{
+		/// Message has been accepted by an outbound channel
+		MessageAccepted(AccountId, MessageNonce),
+	}
+);
+
+decl_error! {
+	pub enum Error for Module<T: Config> {
+		/// Message has an invalid envelope.
+		InvalidEnvelope,
+		/// Message has an unexpected nonce.
+		BadNonce,
+		/// Target application not found.
+		AppNotFound,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Config> for enum Call where origin: T::Origin {
+
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		#[weight = 0]
+		pub fn submit(origin, message: Message) -> DispatchResult {
+			let relayer = ensure_signed(origin)?;
+			// submit message to verifier for verification
+			let log = T::Verifier::verify(&message)?;
+
+			// Decode log into an Envelope
+			let envelope = Envelope::try_from(log).map_err(|_| Error::<T>::InvalidEnvelope)?;
+
+			let channel = BasicInboundChannel::<T>::new(envelope.source);
```

Comment:
source will be the end user that sent the message into the channel i think, not the source contract?

Reply:
I think envelope.source comes from `msg.sender` on the Ethereum side. Prior to this change, there was a single inbound channel (well, one for basic and one for incentivized). This change creates a channel per user. So if we need to have multiple channels per user, we need a different approach.

Reply:
`envelope.source` is currently the contract that sent the message. The end user account is encoded within `envelope.payload` for our current set of apps (eth, erc20, dot)

Reply:
Hmmm, I'm currently looking into this because tests are failing when I use `tx.origin` instead of `msg.sender` in the outbound channel. Is this the reason?

Reply:
hmm, well for incentivizedchannel, we need the source address of the contract that send the message. for basic, we need that too, but in addition we need the tx.origin of the user. so we'll need an extra field in the basic for this.

Reply:
> `envelope.source` is currently the contract that sent the message. The end user account is encoded within `envelope.payload` for our current set of apps (eth, erc20, dot)

So of course we'll need to add another field like `origin` (or some other name) to the envelope to track the end user who sent the message.

Reply:
Sounds good.
We need the mapping to be tx.origin of the user -> nonce then. 
The remaining question then is if this is enough granularity or we want to map (user, channel or contract) -> nonce
If that's the case, we'd need both the tx.origin _and_ the envelope.source.
What do you think?

Reply:
i think we need both

tx.origin -> nonce mapping for replay protection checks
envelope.source for application contract address, so that the application receiving the payload can verify it came from the right contract app